pandas/core/dtypes/cast.py

```diff
@@ -1314,6 +1315,30 @@ def find_result_type(left_dtype: DtypeObj, right: Any) -> DtypeObj:
         #  which will make us upcast too far.
         if lib.is_float(right) and right.is_integer() and left_dtype.kind != "f":
             right = int(right)
+        # After NEP 50, numpy won't inspect Python scalars
+        # TODO: do we need to recreate numpy's inspection logic for floats too
+        # (this breaks some tests)
+        if isinstance(right, int) and not isinstance(right, np.integer):
```

Comment:
I'm the most worried about here.

A lot of things seem to hit around here, but the test coverage isn't great.

We need the modifications here, since result_type won't inspect Python scalars anymore, so we have to figure out the "smallest" dtype for right ourselves.

2 questions here are:
1. Is ``min_scalar_type`` the right way to approach this?
2. Does the "hack" below for deciding whether the "smallest" dtype that can contain value make sense?